include/libassert/platform.hpp

```diff
@@ -0,0 +1,343 @@
+    #define LIBASSERT_HAS_CPP_ATTRIBUTE(x) __has_cpp_attribute(x)
+#else
+    #define LIBASSERT_HAS_CPP_ATTRIBUTE(x) 0
+#endif
+
+
+///
+/// Compiler attribute support.
+///
+
+#if LIBASSERT_HAS_CPP_ATTRIBUTE(nodiscard) && LIBASSERT_STD_VER >= 20
+    #define LIBASSERT_ATTR_NODISCARD_MSG(msg) [[nodiscard(msg)]]
+#else // Assume we have normal C++17 nodiscard support.
+    #define LIBASSERT_ATTR_NODISCARD_MSG(msg) [[nodiscard]]
+#endif
+
+
+#if LIBASSERT_HAS_CPP_ATTRIBUTE(likely)
+    #define LIBASSERT_LIKELY [[likely]]
+#else
+    #define LIBASSERT_ATTR_LIKELY
+#endif
+
+
+#if LIBASSERT_HAS_CPP_ATTRIBUTE(unlikely)
+    #define LIBASSERT_ATTR_UNLIKELY [[unlikely]]
+#else
+    #define LIBASSERT_ATTR_UNLIKELY
+#endif
+
+
+#if LIBASSERT_HAS_CPP_ATTRIBUTE(no_unique_address)
+    #define LIBASSERT_ATTR_NO_UNIQUE_ADDRESS [[no_unique_address]]
+#else
+    #define LIBASSERT_ATTR_NO_UNIQUE_ADDRESS
+#endif
+
+
+///
+/// C++20 feature support.
+///
+
+#if defined(__cpp_conditional_explicit) && __cpp_conditional_explicit >= 201806L
+    #define LIBASSERT_CPP20_EXPLICIT_BOOL(expr) explicit(expr)
+#else
+    #define LIBASSERT_CPP20_EXPLICIT_BOOL(expr) explicit
+#endif
+
+
+#if defined(__cpp_consteval) && __cpp_consteval >= 201811L
+    #define LIBASSERT_CONSTEVAL consteval
+#else
+    #define LIBASSERT_CONSTEVAL constexpr
+#endif
+
+
+#if LIBASSERT_STD_VER >= 20
+    #define LIBASSERT_CPP20_CONSTEXPR constexpr
+#else
+    #define LIBASSERT_CPP20_CONSTEXPR
+#endif
+
+
+///
+/// C++23 feature support.
+///
+
+// Permit static constexpr variables in constexpr functions in C++23.
+#if defined(__cpp_constexpr) && __cpp_constexpr >= 202211L
+    #define LIBASSERT_CONSTEXPR23_STATIC_VAR static constexpr
+#else
+    #define LIBASSERT_CONSTEXPR23_STATIC_VAR constexpr
+#endif
+
+
+#if defined(__cpp_if_consteval) && __cpp_if_consteval >= 202106L
+    #define LIBASSERT_IF_CONSTEVAL if consteval
+#else
+    #define LIBASSERT_IF_CONSTEVAL if constexpr
+#endif
+
+
+#if LIBASSERT_STD_VER >= 23
+    #define LIBASSERT_CONSTEXPR23 constexpr
+#else
+    #define LIBASSERT_CONSTEXPR23
+#endif
+
+///
+/// C++26 feature support.
+/// 
+
+
+// Wrapper macro to allow support for C++26's user generated static_assert messages.
+// The backup message version also allows for the user to provide a backup version that will
+// be used if the compiler does not support user generated messages.
+// More info on user generated static_assert's
+// can be found here: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2741r1.pdf
+#if defined(__cpp_static_assert) && __cpp_static_assert >= 202306L
+    #define LIBASSERT_USER_STATIC_ASSERT(cond, constant) static_assert(cond, constant)
+    #define LIBASSERT_USER_STATIC_ASSERT_BACKUP_MSG(cond, msg, constant) static_assert(cond, constant)
+#else
+    #define LIBASSERT_USER_STATIC_ASSERT(cond, constant) static_assert(cond)
+    #define LIBASSERT_USER_STATIC_ASSERT_BACKUP_MSG(cond, msg, constant) static_assert(cond, msg)
+#endif
```

Comment:
This is exciting :) I'm wondering if this would actually be good to move to the main assert.hpp header since it is more core the library rather than just support.

Reply:
We absolutely could though the name is a bit cumbersome for a library item vs just a support item. What are your thoughts here?

Reply:
Also I think if we were to bring this into the main header that a support function to aid in creating usable formatted messages would be ideal.

You can achieve a compile time formatted message in fmt but I don't know how feasible a helper item is with just std::format?

Here is how it would work with fmt.

```cpp
auto msg = [](int value) {
        char c[100] = {};
        fmt::format_to(c, FMT_COMPILE("The answer is {}."), value);
        return std::string(c);
    };
    ```

Reply:
I agree a helper would be good but we can hold of for now since it's not immediately clear the best way to do it, std::format doesn't have an api like this unfortunately. Happy to go either way on bringing it into the main header now or doing so later.

Reply:
Yeah I agree that for this PR I don't think it should yet be added. I'll for now leave a TODO with the macros definition for future reminders for now.